crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+                update.update.issuer,
+                state.consensus_state.name()
+            )),
+        }
+    }
+}
+
+/// An instance of this resource is created when `JWKManager` starts the QC update building process for an issuer.
+/// Then `JWKManager` needs to hold it. Once this resource is dropped, the corresponding QC update process will be cancelled.
+#[derive(Clone, Debug)]
+pub struct QuorumCertProcessGuard {
+    handle: Option<AbortHandle>,
+}
+
+impl QuorumCertProcessGuard {
+    pub fn new(handle: AbortHandle) -> Self {
+        Self {
+            handle: Some(handle),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn dummy() -> Self {
+        Self { handle: None }
+    }
+}
+
+impl Drop for QuorumCertProcessGuard {
+    fn drop(&mut self) {
+        let QuorumCertProcessGuard { handle } = self;
+        if let Some(handle) = handle {
+            handle.abort();
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum ConsensusState {
+    NotStarted,
+    InProgress {
+        my_proposal: ObservedUpdate,
+        abort_handle_wrapper: QuorumCertProcessGuard,
+    },
+    Finished {
+        vtxn_guard: TxnGuard,
+        my_proposal: ObservedUpdate,
+        quorum_certified: QuorumCertifiedUpdate,
+    },
+}
+
+impl PartialEq for ConsensusState {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (ConsensusState::NotStarted, ConsensusState::NotStarted) => true,
+            (
+                ConsensusState::InProgress {
+                    my_proposal: update_0,
+                    ..
+                },
+                ConsensusState::InProgress {
+                    my_proposal: update_1,
+                    ..
+                },
+            ) if update_0 == update_1 => true,
+            (
+                ConsensusState::Finished {
+                    my_proposal: update_0,
+                    ..
+                },
+                ConsensusState::Finished {
+                    my_proposal: update_1,
+                    ..
+                },
+            ) if update_0 == update_1 => true,
+            _ => false,
+        }
+    }
+}
+
+impl Eq for ConsensusState {}
+
+impl ConsensusState {
+    pub fn name(&self) -> &str {
+        match self {
+            ConsensusState::NotStarted => "NotStarted",
+            ConsensusState::InProgress { .. } => "InProgress",
+            ConsensusState::Finished { .. } => "Finished",
+        }
+    }
+
+    pub fn my_proposal_cloned(&self) -> ObservedUpdate {
+        match self {
+            ConsensusState::InProgress { my_proposal, .. }
+            | ConsensusState::Finished { my_proposal, .. } => my_proposal.clone(),
+            _ => panic!("my_proposal unavailable"),
+        }
+    }
+}
+
+impl Default for ConsensusState {
+    fn default() -> Self {
+        Self::NotStarted
+    }
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct PerProviderState {
+    pub on_chain: Option<ProviderJWKs>,
```

Comment:
When this will be None? The new always takes in onchain state. How is this initialized when there is no onchain state?

Reply:
when we need a default `PerProviderState`, which happens e.g. when we start an epoch with a newly added provider.

Reply:
Yes, but I didn't see it in later PRs. Can you point it to me?

Reply:
![image](https://github.com/aptos-labs/aptos-core/assets/4213569/c1444efe-b930-40a6-b883-27dbb4cf67f5)